### ISSUE TITLE
Make JupyterHub More Configurable

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#56]
What are the common points of JupyterHub that users will want to configure?

  - Default images ?
  - Auth mode?
  - SSL?
      - This is probably blocked on Kubernetes adding good support for KMS (#54)

In addition or alternatively, we could show people how they could easily provide their own KubeSpawner py file to use in their component.

We might want to look at the existing helm package for JupyterHub and think about whether that can be reused in some fashion.

/cc @foxish @yuvipanda

**Comment by yuvipanda** [comment]
Yes, the things we've found that people ask for (and have added to the chart):

1. Different authenticators
2. Automatic HTTPS with Let's encrypt
3. Custom ways to specify how user homedirs are mounted (PVC provisioning)
4. Whitelisting users / setting users as admin users
5. Automatic culling (stop user notebooks after X minutes of inactivity)

Would love to explore how we can use our helm chart here as a dependency.

**Comment by pdmack** [comment]
Yep, I was thinking we would ideally load the "blessed" list of NB images from whenever they are released to gcr.io. @jlewi @ankushagarwal , at some point will we be tagging these to Kubeflow releases instead of the date-sha moving target?

**Comment by jlewi** [comment]
We are now tagging Jupyter images with the release.

**Comment by jlewi** [comment]
Elaborating on the idea of using a ConfigMap. I think we'd like the full ConfigMap/YAML spec to be a ksonnet prototype. The YAML could be specified via jsonnet. 

This way the full spec would be a component inside the user's ksonnet application. So they could easily make changes to it without us needing to expose parameters for every feature. 

Furthermore, can we implement this such that we don't have to add explicit support to kube_spawner.py to support a given option.
http://jupyterhub-kubespawner.readthedocs.io/en/latest/spawner.html#module-kubespawner.spawner

Ideally, we should just be able to list options in the config map corresponding to kube-spawner options and they should be automatically applied.

**Comment by jlewi** [comment]
@pdmack Any interest in picking this up?

**Comment by jlewi** [comment]
See this thread:
https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubeflow-discuss/CKx5wVqzfAs/NoRMdWm9DQAJ

It looks like JupyterHub has an option to cull idle notebooks which would be very useful. The instructions
https://zero-to-jupyterhub.readthedocs.io/en/latest/user-management.html

Indicate it is options specified in the helm chart's config.yaml; so we probably need to look at the helm chart to figure out where the culling option gets set.

@yuvipanda Do you have a link to the source for the JupyterHub helm chart? I was having trouble finding it.

**Comment by clkao** [comment]
The cull service is actually somewhat problematic as it kills browser-inactive users, even though they might have running kernels.

newer jupyter notebook has built in idle culling based on kernel activity, but this needs to be enabled in notebook.  mybinder uses [singleuser extraEnv](https://github.com/jupyterhub/mybinder.org-deploy/blob/master/appendix/extra_notebook_config.py) to configure notebook's MappingKernelManager.

**Comment by jlewi** [comment]
@clkao Thanks; so just to double clarify all we need to do is set the environment variables as specified here: [extra_notebook_config.py](https://github.com/jupyterhub/mybinder.org-deploy/blob/master/appendix/extra_notebook_config.py)? I assume we set this in the pod running Jupyter and not JupyterHub?

**Comment by clkao** [comment]
Let me clarify a bit further since there are multiple things involved:

1. Notebook pod: needs to tweak MappingKernelManager to enable shutdown_no_activity_timeout and kernel culling.
2. Notebook pod: Uses extra_notebook_config to set (1) from env var
3. Hub pod: use singleuser extraEnv to config to configure (2)

(2) is tricky, as you need to bake it into the notebook image which is hard to guarantee if you allow custom images. for images based off docker-stack (jupyter/*), you can actually mount configmap under /usr/local/start-notebook.d/ to provide the env processing

**Comment by pdmack** [comment]
I think it would be useful to focus this issue. As I understand it, the idea would be to load [known kubespawner config](https://jupyterhub-kubespawner.readthedocs.io/en/latest/spawner.html#module-kubespawner.spawner) simply via some kind of YAML/JSON file that could be loaded and read by our spawner Python file. 

The PVC issue for surfacing those objects in the UI is separately tracked at #34 
The culling issue is already at #1394 

So, this issue is solely for the purpose of the E-Z config map approach I described.

Agreed?

**Comment by jlewi** [comment]
@pdmack Agree on the scoping of this issue.

**Comment by jlewi** [comment]
@pdmack feel free to open a new issue to clarify the scope and then close this as a dupe of that.

**Comment by chrisheecho** [comment]
/remove-priority p1
/priority p2

No engineering resource assigned

**Comment by jlewi** [comment]
Closing this out since as of 0.5 JupyterHub is no longer used.
